src/coreclr/src/System.Private.CoreLib/src/System/Threading/Thread.CoreCLR.cs

```diff
@@ -532,6 +544,142 @@ public static int GetCurrentProcessorId()
+        private static bool SimpleCoreIdSpeedCheck()
+        {
+            // warm up the code paths.
+            int id = UninlinedThreadStatic() | GetCurrentProcessorNumber();
+            long _05usec = Stopwatch.Frequency / 1000000;
+
+            // limit quick test to 100 microseconds.
+            // If we are on slow hardware, we should calibrate anyways.
+            long limit = Stopwatch.Frequency / 10000 + Stopwatch.GetTimestamp();
+            for (int i = 0; i < 10; i++)
+            {
+                int iters = 1;
+                long t1 = 0;
+                // double the sample size until it is 0.5 usec.
+                while (t1 < _05usec)
+                {
+                    iters *= 2;
+                    t1 = Stopwatch.GetTimestamp();
+                    for (int j = 0; j < iters; j++)
+                    {
+                        id = GetCurrentProcessorNumber();
+                    }
+                    t1 = Stopwatch.GetTimestamp() - t1;
+                }
+
+                // assuming TLS cannot be a lot slower than getting ID, this should take 1-5 usec
+                long t2 = Stopwatch.GetTimestamp();
+                for (int j = 0; j < iters; j++)
+                {
+                    UninlinedThreadStatic();
+                }
+                long t3 = Stopwatch.GetTimestamp();
+
+                // if getting ID took longer than 2x TLS access, we should consider caching.
+                if (t3 > limit || (t3 - t2) * 2 < t1)
+                {
+                    return false;
+                }
+            }
+
+            // Make sure the result was not negative, which would indicate "Not Supported"
+            return id >= 0;
+        }
+
+        // NoInlining is to make sure JIT does not CSE and to have a better perf proxy for TLS access.
+        // Measuring inlined ThreadStatic in a loop results in underestimates and unnecessary caching.
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static int UninlinedThreadStatic()
+        {
+            return CoreIdCache.t_currentProcessorId;
+        }
+
+        // We must collect multiple samples to account for irregularities caused by GC and context switches.
+        // Why we keep an array of samples and do not adjust as we go:
+        //   We expect that we will adjust the refresh rate down. If we do that early, we may end up forcing all
+        //   the calibration work to happen much sooner. There is no urgency in being calibrated while the app
+        //   is in start-up mode. That would just add to the "rush hour" traffic.
+        private static int sCalibrationToDo;
+        private static int sCalibrationDone;
+        // 25 is chosen to budget the sampling under 100 msec total, assuming 4 msec per sample.
+        private const int CalibrationSamples = 25;
+        private static double[] sCalibrationSamples = new double[CalibrationSamples * 2];
```

Comment:
Should this be in a separate class so that we allocate the array and do the calibration only when really required?

Reply:
The same class can host the `CoreIdCache`.

Reply:
I did not want to have much complexity here. It is a small array allocated once per life of a process. 

I'll check if this can be moved to the nested class and allocation be deferred until needed. If there is something simple we can use that. 
If not, I think this allocation is not a big deal.

---
In reply to: [352952239](https://github.com/dotnet/runtime/pull/467#discussion_r352952239) [](ancestors = 352952239)